### PR TITLE
Feature ehighways resolution unts 318

### LINKED ISSUE #318: Build ehighways resolution units
### What can be improved?

Take regional level NUTS3 data using the correct NUTS year and aggregate up to eHighways resolution. See "data/statistical_units_to_ehighways_regions.csv" in sector-coupled repo for aggregation mapping

### Version

1.2.0.dev

### PR DESCRIPTION
**Pull request by jonathan-peel**
Fixes #318  .

## Checklist

Any checks which are not relevant to the PR can be pre-checked by the PR creator. All others should be checked by the reviewer. You can add extra checklist items here if required by the PR.

- [ ] CHANGELOG updated
- [ ] Minimal workflow tests pass
- [ ] Tests added to cover contribution
- [ ] Documentation updated
- [ ] Configuration schema updated

### CODE REVIEW

**Comment by jnnr** [comment]
A a note: ehighways uses NUTS3 2006 (checked by comparing the region names in anderski_2014 with `statistal-to-ehighways-units.csv)

File: scripts/shapes/units.py
Comment (timtroendle) [inline scripts/shapes/units.py:1]: Many functions in this file lack type hints.

File: config/default.yaml [hunk 1]
```diff
@@ -39,6 +39,9 @@ data-pre-processing:
             MNE: [HRV]
             SRB: [HUN]
             CYP: [ROU]
+    statistical-to-custom-units:
+        # FYI: E-highways were initially defined using NUTS3 units from 2006, but this mapping was manually created to map NUTS units from any year up to 2016 to E-highways. This is picked up by rule `units` to build custom spatial units.
```
Comment (timtroendle) [inline config/default.yaml:43]: This comments seams out of context here. I don't know where this should go, but this looks like the wrong place to me.

File: config/default.yaml [hunk 2]
```diff
@@ -314,6 +317,7 @@ scope:
         first-year: 2016
         final-year: 2016
 shapes: # This config must be consistent with data from https://doi.org/10.5281/zenodo.3244985.
+    # TODO: update this link to Bryn's new one, which includes the potentials for ehighways
```
Comment (timtroendle) [inline config/default.yaml:320]: Please update the link

File: config/default.yaml [hunk 3]
```diff
@@ -423,3 +427,39 @@ shapes: # This config must be consistent with data from https://doi.org/10.5281/
...
+        Sweden: nuts3
+        United Kingdom: nuts3
+        Albania: gadm0
+        Bosnia and Herzegovina: gadm0
+        Macedonia, Republic of: gadm0
+        Montenegro: gadm0
+        Norway: nuts3
+        Serbia: gadm0
+        Switzerland: nuts3
+        Iceland: nuts0
```
Comment (timtroendle) [inline config/default.yaml:465]: Iceland is not currently supported by Euro-Calliope, see #15.

File: config/schema.yaml [hunk 4]
```diff
@@ -141,6 +141,23 @@ properties:
                                     type: string
                                     pattern: ^[A-Z][A-Z][A-Z]$
                                     description: Country to fill missing values from.
+            statistical-to-custom-units:
+                type: object
+                description: >-
+                    Path to a csv file which defines how lower-level statistical units
+                    (i.e., NUTS and GADM) are combined to form units of a custom
```
Comment (timtroendle) [inline config/schema.yaml:148]: GADM are not statistical units.

File: config/schema.yaml [hunk 5]
```diff
@@ -141,6 +141,23 @@ properties:
...
+                    resolutions (e.g., ehighways). FYI: E-highways were initially
```
Comment (timtroendle) [inline config/schema.yaml:149]: The FYI also seems ill-placed here, I am sorry to say.  This is a description of the configuration parameters, but your comment is a comment to the parameter _value_.

File: config/schema.yaml [hunk 6]
```diff
@@ -377,6 +394,10 @@ properties:
                 type: string
                 description: Nuclear technology capacity scenario. "current" will use today's capacities based on the JRC power plant database (PPDB). All other scenario names should point to a configuration CSV file specifying minimum and maximum capacities per country in MW with non-zero capacities (columns=[country, min, max]).
                 pattern: ^(current)|(\w+.csv)$
+            nuts-year:
+                description: Version of the Nomenclature of territorial units for statistics to use in creating spatial units (see https://ec.europa.eu/eurostat/web/nuts/history for more info)
```
Comment (timtroendle) [inline config/schema.yaml:398]: Nomenclature of Territorial Units for Statistics

File: config/schema.yaml [hunk 7]
```diff
@@ -514,6 +535,7 @@ properties:
                                 - "Norway"
                                 - "Serbia"
                                 - "Switzerland"
+                                - "Iceland" # added for ehighways
```
Comment (timtroendle) [inline config/schema.yaml:538]: Please remove. Adding Iceland requires other changes, too.

File: config/schema.yaml [hunk 8]
```diff
@@ -680,10 +702,13 @@ properties:
                         type: string
                         description: Region code
     shapes:
-        description: Source for geospatial shapes of all locations. The config must be consistent with data from https://doi.org/10.5281/zenodo.3244985.
+        description: >-
+            Source for the lowest-level spatial units of all locations.
+            The config must be consistent with data from https://doi.org/10.5281/zenodo.3244985.
+        # TODO: replace zenodo link with Bryn's new one, which includes the potentials for ehighways
```
Comment (timtroendle) [inline config/schema.yaml:708]: Please update link

File: rules/shapes.smk [hunk 9]
```diff
@@ -78,16 +78,16 @@ rule administrative_borders_nuts:
     script: "../scripts/shapes/nuts.py"
 
 
-rule units:
+rule units: # for original resolutions of regional, national, continental
```
Comment (timtroendle) [inline rules/shapes.smk:81]: The comment is incorrect, right? This is for all resolutions.

File: scripts/shapes/units.py [hunk 10]
```diff
@@ -4,25 +4,133 @@
...
 
 def remix_units(
-    path_to_nuts, path_to_gadm, path_to_output, layer_name, layer_config, all_countries
+    path_to_nuts: str,
+    path_to_gadm: str,
+    path_to_output: str,
+    resolution,
+    resolution_config,
+    all_countries,
+    path_to_statistical_to_custom_units,
```
Comment (timtroendle) [inline scripts/shapes/units.py:20]: Last four parameters and output of the function miss type hints.

File: scripts/shapes/units.py [hunk 11]
```diff
@@ -86,12 +182,13 @@ def _write_layer(gdf, path_to_file):
...
-    layer_name = snakemake.wildcards[0]
+    resolution = snakemake.wildcards[0]
     remix_units(
         path_to_nuts=snakemake.input.nuts,
         path_to_gadm=snakemake.input.gadm,
         path_to_output=snakemake.output[0],
-        layer_name=layer_name,
         all_countries=snakemake.params.all_countries,
-        layer_config=snakemake.params.layer_configs[layer_name],
+        resolution_config=snakemake.params.resolution_configs[resolution],
```
Comment (timtroendle) [inline scripts/shapes/units.py:191]: I understand this construct existed before, but it's suboptimal. Please hand over only the configuration of the right resolution to the rule, not the configs of all resolutions.

File: scripts/shapes/units.py [hunk 12]
```diff
@@ -41,24 +149,12 @@ def _validate_source_layers(source_layers):
     ), "Source layers have different crs. They must match."
 
 
-def _validate_layer_config(all_countries, layer_config, layer_name):
-    assert all(country in layer_config for country in all_countries), (
+def _validate_resolution_config(all_countries, resolution_config, layer_name):
+    assert all(country in resolution_config for country in all_countries), (
         f"Layer {layer_name} is not correctly " "defined."
```
Comment (timtroendle) [inline scripts/shapes/units.py:154]: There is a superfluous pair of quotes.

File: scripts/shapes/units.py [hunk 13]
```diff
@@ -4,25 +4,133 @@
...
+    units.loc[units["type"].isnull(), "name"] = "custom_unit"
+    units["type"].fillna("custom_unit", inplace=True)
+
+    return units
+
+
+def _build_layer(resolution_config, source_layers):
+    """
+    Inputs:
+    - resolution_config: a mapping of which statistical unit resolution each country should use as its base
```
Comment (timtroendle) [inline scripts/shapes/units.py:110]: The units are not necessarily statistically. NUTS is, GADM isn't.

File: scripts/shapes/units.py [hunk 14]
```diff
@@ -4,25 +4,133 @@
...
+    - source_layers: a dict with keys as each statistical unit type (nuts[0-3] or gadmn[0-3])
```
Comment (timtroendle) [inline scripts/shapes/units.py:111]: Same here.

File: scripts/shapes/units.py [hunk 15]
```diff
@@ -4,25 +4,133 @@
...
+        ]
+        for country, source_layer in resolution_config.items()
+    ])
+    assert isinstance(layer, pd.DataFrame)
+    return gpd.GeoDataFrame(layer, crs=crs)
 
 
 def _read_source_layers(path_to_nuts, path_to_gadm):
+    """Returns:
+    - source_layers: a dict with keys as each statistical unit type (nuts[0-3] or gadmn[0-3])
```
Comment (timtroendle) [inline scripts/shapes/units.py:131]: GADM isn't statistical.

File: scripts/shapes/units.py [hunk 16]
```diff
@@ -4,25 +4,133 @@
...
-        layer = _continental_layer(layer)
-    _write_layer(layer, path_to_output)
+    _validate_resolution_config(all_countries, resolution_config, resolution)
+    units = _build_layer(resolution_config, source_layers)
+    _validate_layer(units, resolution, all_countries)
+    if resolution == "continental":
+        # sum all of the units together into one block
+        units = _continental_layer(units)
+    elif path_to_statistical_to_custom_units != []:
+        # this code is only executed if there is a [statistical-to-custom-units file corresponding to the chosen resolution] in the config
```
Comment (timtroendle) [inline scripts/shapes/units.py:32]: What are the brackets for?

File: scripts/shapes/units.py [hunk 17]
```diff
@@ -4,25 +4,133 @@
...
+        # sum units together according to custom mapping
+        units = _custom_layer(units, path_to_statistical_to_custom_units)
+    _write_layer(units, path_to_output)
+
+
+def _custom_layer(base_units, path_to_statistical_to_custom_units):
+    """Combine the lower-level statistical units (aspecified in resolution_config)
```
Comment (timtroendle) [inline scripts/shapes/units.py:39]: Typo in comment.

Comment (timtroendle) [reply scripts/shapes/units.py:39]: The units are not necessarily statistical.

Comment (timtroendle) [reply scripts/shapes/units.py:39]: Is it useful to mention `resolution_config` here? It's not part of the function anyway.

File: config/default.yaml [hunk 18]
```diff
@@ -39,6 +39,9 @@ data-pre-processing:
             MNE: [HRV]
             SRB: [HUN]
             CYP: [ROU]
+    statistical-to-custom-units:
```
Comment (timtroendle) [inline config/default.yaml:42]: As mentioned in some other comments, the base units are not necessarily statistical. They are for eHighways, but if you want to be more flexible than eHighways (which you don't have to be, but which you are), then this does not make sense. Why not: `base-to-custom-units`?

File: scripts/shapes/units.py [hunk 19]
```diff
@@ -4,25 +4,133 @@
...
+    into the custom units, according to statistical_to_custom_units, using _merge_units
+    function.
+
+    Inputs:
+    - path_to_statistical_to_custom_units: csv file that specifies which NUTS and GADM
```
Comment (timtroendle) [inline scripts/shapes/units.py:44]: Here again, there should be no `statistical` in the name if you want to be more flexible than eHighways.

File: scripts/shapes/units.py [hunk 20]
```diff
@@ -4,25 +4,133 @@
...
+        units contribute to each custom unit
+    - layer: the GeoDataFrame covering the whole scope, where each country is formed
+        from units corresponding to the resolution specified in resolution_config
+    Outputs:
+    -
```
Comment (timtroendle) [inline scripts/shapes/units.py:49]: Output is missing.

File: scripts/shapes/units.py [hunk 21]
```diff
@@ -4,25 +4,133 @@
...
+
+    Assumes:
+    - the base-level statistical units in layer and are either NUTS3 or whole countries (NUTS0 or GADM0)
```
Comment (timtroendle) [inline scripts/shapes/units.py:52]: GADM is not... you know what. :D

File: config/schema.yaml [hunk 22]
```diff
@@ -377,6 +394,10 @@ properties:
                 type: string
                 description: Nuclear technology capacity scenario. "current" will use today's capacities based on the JRC power plant database (PPDB). All other scenario names should point to a configuration CSV file specifying minimum and maximum capacities per country in MW with non-zero capacities (columns=[country, min, max]).
                 pattern: ^(current)|(\w+.csv)$
+            nuts-year:
```
Comment (timtroendle) [inline config/schema.yaml:397]: This parameter does not exist in the config. Please remove.

File: scripts/shapes/units.py [hunk 23]
```diff
@@ -4,25 +4,133 @@
...
+    - any units which are whole countries will also be whole countries at the custom resolution
+    """
+
+    stat_to_custom_units_df = pd.read_csv(path_to_statistical_to_custom_units, header=0)
+    nuts_year = 2013  # This was a variable in Bryn's SC EC, but hardcoded to 2013 in EC. Consider that the year needs to match for both the (downloaded nuts units) and (nuts to custom region mapping)
```
Comment (timtroendle) [inline scripts/shapes/units.py:57]: Make this a constant at the top of the file.

Comment (timtroendle) [reply scripts/shapes/units.py:57]: I'd remove the first sentence. Alternatively, point to the line in the repo. "Bryn's SC EC" will not make sense to everyone and forever.

**Review by timtroendle — CHANGES_REQUESTED** [review]
Awesome! It looks good, but it requires a few changes, like type hints, documentation, and solving some inconsistencies.

The one thing I am not liking stems from the source: There are now two different ways in which resolutions are defined: the previously-existing `shape` config in the config AND the new `statistical-to-custom-units` file. Ideally, there would be only one place.

I don't think this should be handled within this PR or ticket. But I would like to keep this in mind and consider solving in the future. @brynpickering, do you agree?

**Comment by brynpickering** [comment]
> I don't think this should be handled within this PR or ticket. But I would like to keep this in mind and consider solving in the future. @brynpickering, do you agree?

I don't think it's necessarily a problem. You have standardised boundaries coming from various data sources (nuts, gadm, ...) and they are your building blocks in `shapes`. Then you have a mechanism to merge those shapes together with a custom mapping. It means we know that all shapes in the model are built from standard boundaries and there's no need for a mechanism to injest custom shapes.

**Review by brynpickering — COMMENTED** [review]
Not sure what the issue is with setting a nuts year at the config level and then downloading that nuts year when generating any units (as I did in SC-EC). Since we only use NUTS for `ehighways` at the moment it effectively allows you to choose the NUTS year that will be used to create those shapes.

We'll need the ability at a later point to download arbitrary NUTS years (for subnationalising using Eurostat datasets) so I guess we can leave it until then to introduce it as a feature.

**Pull request merged**